File: product-service/src/main/java/com/heulwen/productservice/service/ProductService.java
package com.heulwen.productservice.service;

import com.heulwen.productservice.api.request.ProductRequest;
import com.heulwen.productservice.api.response.ProductResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductService {
    ProductResponse createProduct(ProductRequest request);
    ProductResponse updateProduct(Long id, ProductRequest request);
    void deleteProduct(Long id);
    ProductResponse getProduct(Long id);
    Page<ProductResponse> getProductsForUser(String search, Long categoryId, Pageable pageable);
    Page<ProductResponse> getProductsForAdmin(String search, Long categoryId, Pageable pageable);
    void reduceStock(Long id, int quantity);
}

File: frontend-service/src/services/productService.ts
import { apiFetch } from './api';
import { Product, ProductRequest } from '../types/product';
import { PageResponse } from '../types/common';

export const productService = {
  async getProductsForAdmin(search?: string, categoryId?: number, page = 0, size = 10): Promise<PageResponse<Product>> {
    return apiFetch<PageResponse<Product>>('/api/products/admin', {
      method: 'GET',
      params: { search, categoryId, page, size },
    });
  },

  async getProducts(search?: string, categoryId?: number, page = 0, size = 10): Promise<PageResponse<Product>> {
    return apiFetch<PageResponse<Product>>('/api/products', {
      method: 'GET',
      params: { search, categoryId, page, size },
    });
  },

  async getProduct(id: number): Promise<Product> {
    return apiFetch<Product>(`/api/products/${id}`, {
      method: 'GET',
    });
  },

  async createProduct(request: ProductRequest): Promise<Product> {
    return apiFetch<Product>('/api/products', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  },

  async updateProduct(id: number, request: ProductRequest): Promise<Product> {
    return apiFetch<Product>(`/api/products/${id}`, {
      method: 'PUT',
      body: JSON.stringify(request),
    });
  },

  async deleteProduct(id: number): Promise<void> {
    return apiFetch<void>(`/api/products/${id}`, {
      method: 'DELETE',
    });
  }
};
